File: platform/client/src/utils/permissions.ts
import { useAuth } from '@/context/AuthContext';

export type UserRole = 'OWNER' | 'ADMIN' | 'CREATOR' | 'VIEWER';
export type WorkspaceType = 'MAIN' | 'FRANCHISE';

export type Action = 
    | 'workspace:rename'
    | 'workspace:delete'
    | 'workspace:transfer'
    | 'workspace:view'
    | 'team:invite'
    | 'team:manage_roles'
    | 'team:view'
    | 'billing:manage'
    | 'billing:view'
    | 'domains:add'
    | 'domains:verify'
    | 'domains:delete'
    | 'domains:view'
    | 'campaign:create'
    | 'campaign:edit'
    | 'campaign:send'
    | 'campaign:manage'
    | 'campaign:view'
    | 'contacts:import'
    | 'contacts:export'
    | 'contacts:view'
    | 'analytics:view'
    | 'settings:view'
    | 'settings:update'
    | 'api_keys:manage'
    | 'franchise:manage'
    | 'template:view'
    | 'template:manage'
    | 'sender:manage'
    | 'settings:manage';


interface UserContext {
    role: UserRole;
    workspaceType: WorkspaceType;
}

/**
 * Production-grade RBAC permission check for UI elements.
 */
export function can(user: UserContext | null | undefined, action: Action): boolean {
    if (!user || !user.role || !user.workspaceType) {
        return false;
    }

    const { role, workspaceType } = user;

    // 1. WORKSPACE ISOLATION: Franchise workspaces cannot delete infrastructure
    if (workspaceType === 'FRANCHISE') {
        if (['workspace:delete', 'workspace:transfer', 'domains:delete'].includes(action)) {
            return false;
        }
    }

    // 2. ROLE-BASED ACCESS
    switch (role) {
        case 'OWNER':
            return true;

        case 'ADMIN':
            // Admins can do everything EXCEPT critical owner actions
            const ownerOnlyActions = [
                'workspace:rename', 'workspace:delete', 'workspace:transfer',
                'team:manage_roles', 'billing:manage', 'domains:delete', 
                'domains:add', 'domains:verify', 'api_keys:manage',
                'franchise:manage'
            ];
            return !ownerOnlyActions.includes(action);

        case 'CREATOR':
            // Creators focus on content production
            const creatorActions = [
                'campaign:create', 'campaign:edit', 'campaign:view',
                'contacts:import', 'contacts:view', 'analytics:view',
                'team:view', 'domains:view', 'billing:view',
                'template:view', 'template:manage',
                'workspace:view', 'settings:view', 'settings:update'
            ];
            // Explicitly deny manage permissions even if not in ownerOnlyActions
            if (['billing:manage', 'settings:manage'].includes(action)) return false;
            return creatorActions.includes(action as any);

        case 'VIEWER':
            // Viewers are read-only
            const viewerActions = [
                'campaign:view', 'analytics:view', 'contacts:view',
                'team:view', 'domains:view', 'billing:view',
                'template:view', 'workspace:view', 'settings:view'
            ];
            return viewerActions.includes(action as any);

        default:
            return false;
    }
}

File: platform/api/utils/permissions.py
from fastapi import Depends, HTTPException, status
from utils.jwt_middleware import require_active_tenant, JWTPayload, verify_jwt_token

# -----------------------------------
# ROLE PERMISSION MATRIX
# -----------------------------------
# Precise mapping based on production SaaS requirements
ROLE_PERMISSIONS = {
    "OWNER": [
        # Absolute Power
        "workspace:rename", "workspace:delete", "workspace:transfer",
        "team:manage_roles", "billing:manage", "domains:delete", "api_keys:manage",
        # Inherit all below
        "team:invite", "campaign:send", "campaign:manage", "contacts:export",
        "domains:add", "domains:verify", "settings:update", "franchise:manage",
        "campaign:create", "campaign:edit", "contacts:import", "contacts:view",
        "analytics:view", "campaign:view", "team:view", "domains:view", 
        "sender:manage", "template:view", "template:manage", "billing:view",
        "settings:manage", "workspace:view", "settings:view", "asset:create"
    ],
    "ADMIN": [
        "team:invite", "campaign:send", "campaign:manage", "contacts:export",
        "settings:update", "franchise:manage",
        "campaign:create", "campaign:edit", "contacts:import", "contacts:view",
        "analytics:view", "campaign:view", "team:view", "domains:view",
        "sender:manage", "template:view", "template:manage", "billing:view",
        "settings:manage", "workspace:view", "settings:view", "asset:create"
    ],
    "CREATOR": [
        "campaign:create", "campaign:edit", "contacts:import", "contacts:view",
        "analytics:view", "campaign:view", "team:view", "domains:view",
        "template:view", "template:manage",
        "workspace:view", "settings:view", "billing:view", "settings:update", "asset:create"
    ],
    "VIEWER": [
        "analytics:view", "campaign:view", "team:view", "domains:view", "template:view", "billing:view",
        "workspace:view", "settings:view"
    ]
}

def can(payload: JWTPayload, action: str) -> bool:
    """
    Core backend RBAC validator.
    Checks if the user role in the payload has the required action.
    """
    role = getattr(payload, "role", "VIEWER").upper()
    workspace_type = payload.workspace_type
    
    # 1. Workspace Isolation Overrides
    if workspace_type == "FRANCHISE":
        # Production Hardening: Franchises cannot manage infrastructure, billing, or nested franchises.
        # These are exclusively Main Workspace (Parent) administrative concerns.
        if action in [
            "domains:delete", "domains:add", "domains:verify", 
            "workspace:delete", "workspace:transfer", "workspace:rename",
            "billing:manage", "franchise:manage"
        ]:
            return False

    # 2. Permission Check
    permissions = ROLE_PERMISSIONS.get(role, [])
    if action in permissions:
        return True
        
    return False

def require_permission(action: str):
    """
    Dependency generator for FastAPI endpoints.
    Usage: Depends(require_permission('campaign:send'))
    """
    def permission_checker(jwt_payload: JWTPayload = Depends(verify_jwt_token)):
        if not can(jwt_payload, action):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Missing required permission: {action}"
            )
        return jwt_payload
    return permission_checker
